src/raftstore/store/worker/apply.rs

```diff
@@ -617,71 +675,69 @@ impl ApplyDelegate {
             });
     }
 
-    fn handle_raft_entry_normal(
-        &mut self,
-        apply_ctx: &mut ApplyContext,
+    fn handle_raft_entry(
+        mut self,
+        mut apply_ctx: ApplyContext,
         entry: Entry,
-    ) -> Option<ExecResult> {
+    ) -> Box<Future<Item = RaftEntryFutureItem, Error = FutureError> + Send> {
         let index = entry.get_index();
         let term = entry.get_term();
-        let data = entry.get_data();
-
-        if !data.is_empty() {
-            let cmd = util::parse_data_at(data, index, &self.tag);
-
-            if should_write_to_engine(&cmd, apply_ctx.wb().count()) {
-                apply_ctx.commit(self);
-            }
+        match entry.get_entry_type() {
+            EntryType::EntryNormal => {
+                let data = entry.get_data();
+                if !data.is_empty() {
+                    let cmd = util::parse_data_at(data, index, &self.tag);
+
+                    if should_write_to_engine(&cmd, apply_ctx.wb().count()) {
+                        apply_ctx.commit(&mut self);
+                    }
 
-            return self.process_raft_cmd(apply_ctx, index, term, cmd);
-        }
+                    return box self.process_raft_cmd(apply_ctx, index, term, cmd);
+                }
 
-        // when a peer become leader, it will send an empty entry.
-        let mut state = self.apply_state.clone();
-        state.set_applied_index(index);
-        self.apply_state = state;
-        self.applied_index_term = term;
-        assert!(term > 0);
-        while let Some(mut cmd) = self.pending_cmds.pop_normal(term - 1) {
-            // apprently, all the callbacks whose term is less than entry's term are stale.
-            apply_ctx
-                .cbs
-                .last_mut()
-                .unwrap()
-                .push(cmd.cb.take(), cmd_resp::err_resp(Error::StaleCommand, term));
-        }
-        None
-    }
+                // when a peer become leader, it will send an empty entry.
+                let mut state = self.apply_state.clone();
+                state.set_applied_index(index);
+                self.apply_state = state;
+                self.applied_index_term = term;
+                assert!(term > 0);
+                while let Some(mut cmd) = self.pending_cmds.pop_normal(term - 1) {
+                    // apprently, all the callbacks whose term is less than entry's term are stale.
```

Comment:
Could you add more comments to explain why there could be pending proposals with less term?

Reply:
We propose something but not committed when this peer is leader, and then the leader transfer to another peer, this will cause there are pending proposals with less term.